src/main/java/picard/sam/CreateSequenceDictionary.java

```diff
@@ -134,33 +132,53 @@ protected int doWork() {
         }
-        final SAMSequenceDictionary sequences = makeSequenceDictionary(REFERENCE);
-        final SAMFileHeader samHeader = new SAMFileHeader();
-        samHeader.setSequenceDictionary(sequences);
-        final SAMFileWriter samWriter = new SAMFileWriterFactory().makeSAMWriter(samHeader, false, OUTPUT);
-        samWriter.close();
-        return 0;
-    }
 
-    /**
-     * Read all the sequences from the given reference file, and convert into SAMSequenceRecords
-     * @param referenceFile fasta or fasta.gz
-     * @return SAMSequenceRecords containing info from the fasta, plus from cmd-line arguments.
-     */
-    public SAMSequenceDictionary makeSequenceDictionary(final File referenceFile) {
-        final ReferenceSequenceFile refSeqFile =
-                ReferenceSequenceFileFactory.getReferenceSequenceFile(referenceFile, TRUNCATE_NAMES_AT_WHITESPACE);
-        ReferenceSequence refSeq;
-        final List<SAMSequenceRecord> ret = new ArrayList<SAMSequenceRecord>();
-        final Set<String> sequenceNames = new HashSet<String>();
-        for (int numSequences = 0; numSequences < NUM_SEQUENCES && (refSeq = refSeqFile.nextSequence()) != null; ++numSequences) {
-            if (sequenceNames.contains(refSeq.getName())) {
-                throw new PicardException("Sequence name appears more than once in reference: " + refSeq.getName());
+        // SortingCollection is used to check uniqueness of sequence names
+        final SortingCollection<String> sequenceNames = makeSortingCollection();
+        try (BufferedWriter writer = makeWriter()) {
+            final ReferenceSequenceFile refSeqFile = ReferenceSequenceFileFactory.
+                    getReferenceSequenceFile(REFERENCE, TRUNCATE_NAMES_AT_WHITESPACE);
+            SAMSequenceDictionaryCodec samDictCodec = new SAMSequenceDictionaryCodec(writer);
+
+            samDictCodec.encodeHeaderLine(false);
+            // read reference sequence one by one and write its metadata
+            ReferenceSequence refSeq;
+            while ((refSeq = refSeqFile.nextSequence()) != null) {
+                final SAMSequenceRecord samSequenceRecord = makeSequenceRecord(refSeq);
+                samDictCodec.encodeSequenceRecord(samSequenceRecord);
+                sequenceNames.add(refSeq.getName());
+            }
+        } catch (FileNotFoundException e) {
+            throw new PicardException("File " + OUTPUT.getAbsolutePath() + " not found");
+        } catch (IOException e) {
+            throw new PicardException("Can't write to or close output file " + OUTPUT.getAbsolutePath());
+        }
+
+        // check uniqueness of sequences names
+        final CloseableIterator<String> iterator = sequenceNames.iterator();
+        String first = iterator.hasNext() ? iterator.next() : "";
+        while (iterator.hasNext()) {
+            final String next = iterator.next();
+            if (first.equals(next)) {
+                OUTPUT.delete();
```

Comment:
I think it makes more sense to write a temporary file and move it into place when successful than to write a file and then possibly delete it. For example, you might not have permissions to delete files...